File: cp_rust/src/bin/cses-1094-increasing-array.rs
// Created by Ayush Biswas at 2025/07/09 22:20
// https://cses.fi/problemset/task/1094
use cp_lib::*;

// @code begin
use cpio::*;

sol! {
    fn solution(
        _n: usize,
        a: [usize]
    ) -> usize {
        a
            .into_iter()
            .scan(0, |max_seen, ai| {
                *max_seen = (*max_seen).max(ai);
                Some(*max_seen - ai)
            })
            .sum()
    }
}

// @code end

File: cp_haskell/bin/cses-1094-increasing-array.hs
-- Created by Ayush Biswas at 2026/03/03 15:46
-- https://cses.fi/problemset/task/1094

-- @code begin

main :: IO ()
main = do
  n :: Int <- read <$> getLine
  xs :: [Int] <- map read . words <$> getLine
  let (_, res) = foldl nextMove (0, 0) xs
  print res
  where
    nextMove (curr, moves) e =
      let mov = max (curr - e) 0
       in (e + mov, moves + mov)

-- @code end